File: 2101-last-day-where-you-can-still-cross/last-day-where-you-can-still-cross.py
class Solution:
    def latestDayToCross(self, row: int, col: int, cells: List[List[int]]) -> int:
        def find(x):
            if p[x] != x:
                p[x] = find(p[x])
            return p[x]
        
        def union(a, b):
            ap = find(a)
            bp = find(b)
            p[ap] = bp
        
        dx = [-1, -1, 0, 1, 1, 1, 0, -1]
        dy = [0, -1, -1, -1, 0, 1, 1, 1]
        
        p = [i for i in range(row * col + 2)]
        board = [[0 for _ in range(col)] for _ in range(row)]
       
        
        for day, (a, b) in enumerate(cells):
            y, x = a - 1, b - 1
            board[y][x] = 1
            i = y * col + x + 1
            for d in range(8):
                nx = x + dx[d]
                ny = y + dy[d]
                if not (0 <= nx < col and 0 <= ny < row):
                    continue
                if board[ny][nx] == 0:
                    continue
                j = ny * col + nx + 1
                union(i, j)
                
            if x == 0:
                union(0, i)
            elif x == col - 1:
                union(i, row * col + 1)
                
            if find(0) == find(row * col + 1):
                return day

File: 2101-last-day-where-you-can-still-cross/last-day-where-you-can-still-cross.js
/**
 * @param {number} row
 * @param {number} col
 * @param {number[][]} cells
 * @return {number}
 */
var latestDayToCross = function(row, col, cells) {
    const find = (x) => {
        if (p[x] !== x) p[x] = find(p[x]);
        return p[x];
    }

    const union = (a, b) => {
        const ap = find(a);
        const bp = find(b);
        p[bp] = ap;
    }
    
    const n = row * col + 2;
    const board = Array.from({length: row}, () => Array.from({length: col}, () => 0));
    const p = Array.from({length:n}, (_, i) => i);

    const dx = [-1, -1, 0, 1, 1, 1, 0, -1];
    const dy = [0, 1, 1, 1, 0, -1, -1, -1];

    for (let i = 0; i < cells.length; i++) {
        const [y, x] = cells[i].map(num => num - 1);
        board[y][x] = 1;
        const now = (y * col) + x + 1; 
        for (let j = 0; j < 8; j++) {
            const nx = x + dx[j];
            const ny = y + dy[j];

            if (!(0 <= nx && nx < col && 0 <= ny && ny < row)) continue;
            if (board[ny][nx] === 0) continue;

            const neighbor = (ny * col) + nx + 1;

            union(now, neighbor)
        }
        if (x === 0) union(0, now)
        if (x === col - 1) union(n - 1, now) 
        if (find(0) === find(n - 1)) return i;
    }
};